superduperdb/vector_search/update_tasks.py

```diff
@@ -40,48 +41,49 @@ def copy_vectors(
-    docs = db.select(select)
-    docs = [doc.unpack() for doc in docs]
-    key = vi.indexing_listener.key
-    if '_outputs.' in key:
-        key = key.split('.')[1]
-    # TODO: Refactor the below logic
-    vectors = []
-    if isinstance(db.databackend, MongoDataBackend):
-        vectors = [
-            {
-                'vector': MongoStyleDict(doc)[
-                    f'_outputs.{vi.indexing_listener.predict_id}'
-                ],
-                'id': str(doc['_id']),
-            }
-            for doc in docs
-        ]
-    elif isinstance(db.databackend, IbisDataBackend):
-        docs = db.execute(select.outputs(vi.indexing_listener.predict_id))
-        from superduperdb.backends.ibis.data_backend import INPUT_KEY
+    if CFG.cluster.vector_search.type != 'pg_vector':
+        vi = db.vector_indices[vector_index]
+        if isinstance(query, dict):
+            # ruff: noqa: E501
+            query: CompoundSelect = Serializable.decode(query)  # type: ignore[no-redef]
+        assert isinstance(query, CompoundSelect)
+        if not ids:
+            select = query
+        else:
+            select = query.select_using_ids(ids)
+        docs = db.select(select)
+        docs = [doc.unpack() for doc in docs]
+        key = vi.indexing_listener.key
+        if '_outputs.' in key:
+            key = key.split('.')[1]
+        # TODO: Refactor the below logic
+        vectors = []
+        if isinstance(db.databackend, MongoDataBackend):
+            vectors = [
+                {
+                    'vector': MongoStyleDict(doc)[
+                        f'_outputs.{vi.indexing_listener.predict_id}'
+                    ],
+                    'id': str(doc['_id']),
+                }
+                for doc in docs
+            ]
+        elif isinstance(db.databackend, IbisDataBackend):
+            docs = db.execute(select.outputs(vi.indexing_listener.predict_id))
+            from superduperdb.backends.ibis.data_backend import INPUT_KEY
 
-        vectors = [
-            {
-                'vector': doc[f'_outputs.{vi.indexing_listener.predict_id}'],
-                'id': str(doc[INPUT_KEY]),
-            }
-            for doc in docs
-        ]
-    for r in vectors:
-        if hasattr(r['vector'], 'numpy'):
-            r['vector'] = r['vector'].numpy()
+            vectors = [
+                {
+                    'vector': doc[f'_outputs.{vi.indexing_listener.predict_id}'],
+                    'id': str(doc[INPUT_KEY]),
+                }
+                for doc in docs
+            ]
+        for r in vectors:
+            if hasattr(r['vector'], 'numpy'):
+                r['vector'] = r['vector'].numpy()
 
-    if vectors:
-        db.fast_vector_searchers[vi.identifier].add(
-            [VectorItem(**vector) for vector in vectors]
-        )
+        if vectors:
+            db.fast_vector_searchers[vi.identifier].add(
+                [VectorItem(**vector) for vector in vectors]
+            )
```

Comment:
The same here

Reply:
@jieguangzhou Remember we discussed in meeting `Ibis` is not compatible with `pgvector` whereas `psycopg2` supports.  It's throwing error when `Ibis` is  trying to access table with vector embeddings created with `psycopg2`.

Reply:
We should have two tables when using the `pg vector` search
1. The model output table(`table1`), all the embedding vectors results are saved here
2. the pg vector search table(`table2`), save the vectors, and build an index here

The workflow for building a vector search index is as follows:

1. Query the vectors from `table1`
2. Use these vectors to build the vector search(`the add method of PostgresVectorSearcher`)

As you can see here, we get the `vectors` and call the add `db.fast_vector_searchers[vi.identifier].add`

Reply:
Hi, you want me to create 2 tables 

1. `outputs.listener1::0` which is created by `ibis`.
2.  `outputs.listener1::0_pgvector` which will be created by `psycopg2`. with the `add method of PostgresVectorSearcher`.

Reply:
No, just one table, the pgvector table created by psycopg2

The first one created by ibis data_backend automatically

Reply:
@jieguangzhou I mean the same thing.

1. `outputs.listener1::0` which is automatically created by `ibis databackend`.
2. `outputs.listener1::0_pgvector` which will be created by `psycopg2`. 

Just want to check in so that we are on the same page.

Reply:
Yes, we mean the same thing.

Reply:
@jieguangzhou why do we need 2 tables. Why can't we just use the original listener table?

Reply:
This needs to be viewed in conjunction with comment https://github.com/SuperDuperDB/superduperdb/pull/1926#discussion_r1570648286 ; here are my thoughts:

1. The saving behavior of model outputs should not be limited by downstream applications (such as vector search). If we need to compatible vector search separately, it would increase the complexity of saving query results. This should be a very pure interface. If vector search is to be built, then an index should be built on this table on the vector search side (if an index cannot be built, a separate table needs to be created).
2. The vector search component should be independent. Ideally, we should be able to switch the underlying engine of my vector search at any time. If we couple vector search with saving, then expandability will decrease. If my vector search uses two different Postgres databases from databackend (one for data saving and one for vector search calculations), it cannot be supported.
3. Ideally, if the PostgresVectorSearcher can determine that the uri being used is the same as the databackend's uri, it should first attempt to build an index in the model output table. If building an index is not possible, then a separate table needs to be created.


WDYT? @blythed

Reply:
I understand that developers could potentially use pg_vector as an external vector-store. However, most likely 
users will be postgres users. 

By hosting the vector-search with pg_vector, we avoid the problem of mapping the data to a new table/ database. This is exactly like MongoDB Atlas vector-search.

Reply:
I just conducted a basic test with pgvector:
1. If a vector table is created using pgvector, ibis currently cannot directly read the data inside, as it throws an error for not being able to parse the vector type data of pgvector. Adaptations are necessary; otherwise, our data backend will no longer be able to access this table. I suspect that it would require activating the vector feature specifically for the ibis backend’s connection, which could be quite troublesome.
2. To create and access vector type data in pgvector, each connection needs to activate the vector feature to enable data interchange.
It is suggested to split it into two tables, as ibis does not seem to be well compatible with pgvector at the moment.
If merging into one table is considered, the following tests need to be conducted:
1. Test with two types of SQL databases (Postgres and SQLite) to ensure that non-Postgres vector implementations are compatible and do not affect the original SQL functionalities. Since SQLite already has unit tests, only Postgres needs to be tested.
2. Conduct application tests in cases where the vector search type is specified as pgvector, including vector searches and non-vector searches, to ensure that non-vector search functions are preserved normally. These tests should be integration tests

@makkarss929 @blythed